WIP Timeseries base

mop9047 opened the pull request:
Adds a TimeSeries Feature so sequences of data can be used in ML5

This version extends from the Neural Network class with some overrides to allow sequences of data. It also includes some helper functions to simplify loading timeseries data.

Key Features:

- `addData` in the format of Array of Objects [{x: ,y: },{x: ,y: }] or Array of Arrays [[,],[,]]
- `addDefaultLayers` which uses `tsLayers.js` file to define the architectures, based on `task` and `dataMode` (still needs to be discussed).
- `padCoordinates(sequence, targetLength)` uses RDP algorithm to simplify coordinate data to a specified target length
- `slidingWindow(data, featureKeys, targetKeys, batchLength = null)` and `sampleWindow(data)` uses sliding Window algorithm to create batches from sequence data and take a single sample for prediction

gohai commented on src/TimeSeries/index.js:
```diff
@@ -0,0 +1,193 @@
+
+  async init() {
+    // workaround for Error
+    setBackend("webgl");
```
Regular NeuralNetwork leaves this to the sketch to set, so I'd suggest we do the same. (Or, change NN in the same manner. I don't recall, but there might have been previous discussion why we didn't transparently forced the backend to "webgl" in such a way.)

mop9047 replied:
I agree, I'll leave this comment unresolved in this case, so it doesn't get forgotten

gohai commented on src/TimeSeries/index.js:
```diff
@@ -0,0 +1,193 @@
+        return this.createNetworkLayers(layers);
+
+      default:
+        console.log("no inputUnits or outputUnits defined");
```
console.warn(), console.error()?

mop9047 replied:
done :) I chose to do console.warn instead as it's not necessarily an error and the layers are chosen automatically

gohai commented on src/TimeSeries/index.js:
```diff
@@ -0,0 +1,193 @@
+        optimizer: tf.train.adam,
+        metrics: ["accuracy"],
+      };
+    }
```
Maybe a catch-all else that informs the user that they specified an unknown/unsupported task?

mop9047 replied:
I agree! these were a copied block from the DIY NN class, I will add a catch-all assuming that regression is the default, although maybe there is a discussion needed to decide otherwise. For the DIY NN class, I see that the default task is set to null. I could also add a condition to check if the task remains null and send an error instead.

gohai commented on src/TimeSeries/timeSeriesData.js:
```diff
@@ -0,0 +1,381 @@
+   *
+   * @return {{ inputs: tf.Tensor, outputs: tf.Tensor }}
+   */
+  // eslint-disable-next-line class-methods-use-this, no-unused-vars
```
Is this needed (for our prettifier)?

mop9047 replied:
this line was in the DIY NN class, I tested without it and there nothing changed or any errors so I will probably remove.

gohai commented on examples/timeSeries-hand-gestures/sketch.js:
```diff
@@ -0,0 +1,192 @@
+}
+
+function whileTraining(epoch) {
+  console.log(epoch);
```
Easy to get confused with console.log()s that aren't your (the user's) own... :)

Best to make them as descriptive as possible (e.g. `console.log("In epoch", epoch);` and then commenting them out to, so that they're still available for future debugging.

gohai commented on src/TimeSeries/timeSeriesData.js:
```diff
@@ -0,0 +1,381 @@
+      const batch = normalized[features[0]].length / seriesStep;
+
+      this.meta.seriesShape = [seriesStep, feature_length];
+      console.log("series shape", this.meta.seriesShape);
```
Easy to get confused with console.log()s that aren't your (the user's) own... :)

Best to make them as descriptive as possible (e.g. console.log("In epoch", epoch); and then commenting them out to, so that they're still available for future debugging.

mop9047 replied:
Just removed this comment as it was only helpful for a specific error that I already fixed.

gohai commented on src/TimeSeries/timeSeriesData.js:
```diff
@@ -0,0 +1,381 @@
+
+    const sample = this.data.raw[0];
+
+    //consistent dTypes have already been checked at add data
```
Nitpick: space after `//`

mop9047 replied:
done :)

gohai commented on src/TimeSeries/timeSeriesData.js:
```diff
@@ -0,0 +1,381 @@
+    const sample = this.data.raw[0];
+
+    //consistent dTypes have already been checked at add data
+    const xs = Object.keys(sample.xs[0]); //since time series data is in form of array
```
Nitpick: space after `//`

gohai commented on examples/timeSeries-hand-gestures/sketch.js:
```diff
@@ -0,0 +1,192 @@
+  let options = {
+    outputs: ["label"],
+    task: "classification",
+    dataMode: "spatial",
```
If my memory serves me right: I thought the plan was to drop the extra property to activate the special handling for spatial data, and instead fold it into `task` (similarly to `classification` and `imageClassification`)?

gohai commented on examples/timeSeries-load-model-hand-gestures/model/model.json:
```diff
@@ -0,0 +1 @@
+{"modelTopology":{"class_name":"Sequential","config":{"name":"sequential_1","layers":[{"class_name":"Conv1D","config":{"filters":8,"kernel_initializer":{"class_name":"VarianceScaling","config":{"scale":1,"mode":"fan_avg","distribution":"normal","seed":null}},"kernel_regularizer":null,"kernel_constraint":null,"kernel_size":[3],"strides":[1],"padding":"valid","dilation_rate":[1],"activation":"relu" ...
```
Nitpick: adding a newline at the end of the file makes Git happy

mop9047 replied:
done :)

gohai commented on examples/timeSeries-weather-prediction/sketch.js:
```diff
@@ -0,0 +1,183 @@
+ * Learn more about the ml5.js project: https://ml5js.org/
+ * ml5.js license and Code of Conduct: https://github.com/ml5js/ml5-next-gen/blob/main/LICENSE.md
+ *
+ * This example demonstrates Training a Stock Price Predictor through ml5.TimeSeries.
```
(This still refers to stock prices, fyi)

gohai commented on examples/timeSeries-hand-gestures/sketch.js:
```diff
@@ -0,0 +1,192 @@
+    task: "classification",
+    dataMode: "spatial",
+    debug: "true",
+    learningRate: 0.001,
```
A comment explaining whether/why deviating from the default was necessary here, and how came up with this value, might be instructive. (e.g. "the default learning rate of 0.01 didn't converge with this data set - lowering the ...")

gohai commented on examples/timeSeries-hand-gestures/sketch.js:
```diff
@@ -0,0 +1,192 @@
+      // once sequence reaches the seqLength, add sequence as just one X value
+    } else if (sequence.length > 0) {
+      // get the training word from the input box
+      let train_word = nameField.value();
```
Stylistic nitpick: perhaps `trainWord`, since the rest of the example uses camelCase

gohai commented on src/TimeSeries/index.js:
```diff
@@ -0,0 +1,193 @@
+    if (typeof inputs === "number") {
+      inputShape = inputs;
+    } else if (Array.isArray(inputs) && inputs.length > 0) {
+      inputShape = inputs.length; //will be fed into the tensors later
```
Nitpick: space after `//`

mop9047 replied:
done :)

shiffman commented on examples/timeSeries-hand-gestures/sketch.js:
```diff
@@ -0,0 +1,192 @@
+
+  let counts = "";
+  let keys = Object.keys(trainingWords);
+  console.log("keys", keys);
```
Just noting this code which includes `Object.keys()` and a `for in` loop is not typical of p5.js style and may be confusing to beginners. They may be necessary in which case I might suggest explanatory comments. Also, I would probably remove the console logs from the examples.

mop9047 replied:
I have removed the console logs! and for this one I simplified the example even further, in the latest commit I refactored the example code to simply use "gesture #1" and "gesture #2" instead. I also made the controls similar to the Mouse Gesture example to keep things consistent and more simple. Let me know if you still want to keep the user inputted ones.

shiffman commented on examples/timeSeries-weather-prediction/sketch.js:
```diff
@@ -0,0 +1,183 @@
+    dataMode: "linear",
+    debug: "true",
+    learningRate: 0.01,
+    output: ["label"],
```
I'm confused by this, it doesn't match up to me with the format of the data? Also, ml5.js will infer inputs and outputs from the data itself so maybe unecessary to include?

mop9047 replied:
You're right! it was unnecessary. However, for this example, I changed the output to be the same variable targets which acts both as defining which will be included as the outputs and in the sliding window helper function. However in this case, I used all features as targets to create a loop for creating continuous data.

shiffman reviewed (commented):
Some additional comments along the lines of @gohai's review!  Also, I see this PR still includes `ml5.timeSeries()`, my memory is shaky of course, but I recall a discussion around this implementation being a part of `ml5.neuralNetwork()` with a new "time series" task? Is that correct?

gohai commented:
> Some additional comments along the lines of @gohai's review! Also, I see this PR still includes `ml5.timeSeries()`, my memory is shaky of course, but I recall a discussion around this implementation being a part of `ml5.neuralNetwork()` with a new "time series" task? Is that correct?

My memory is also shaky, @shiffman - but I like the idea. Since the options are being passed to the factory method, we could _keep_ the inheritance model that Mathew built, and have `ml5.neuralNetwork()` simply instantiate the specialized class whenever encounters a related `task` option. The alternative is to integrate Mathew's work even more tightly into the existing NeuralNetwork classes. Do you have a preference @mop9047?

mop9047 commented on examples/timeSeries-weather-prediction/sketch.js:
```diff
@@ -0,0 +1,183 @@
+    task: "regression",
+    dataMode: "linear",
+    debug: "true",
+    learningRate: 0.01,
```
done :)

mop9047 commented:
> > Some additional comments along the lines of @gohai's review! Also, I see this PR still includes `ml5.timeSeries()`, my memory is shaky of course, but I recall a discussion around this implementation being a part of `ml5.neuralNetwork()` with a new "time series" task? Is that correct?
> 
> My memory is also shaky, @shiffman - but I like the idea. Since the options are being passed to the factory method, we could _keep_ the inheritance model that Mathew built, and have `ml5.neuralNetwork()` simply instantiate the specialized class whenever encounters a related `task` option. The alternative is to integrate Mathew's work even more tightly into the existing NeuralNetwork classes. Do you have a preference @mop9047?

Yes! I do remember having that discussion, but I don't clearly remember if we had a decision on which method to do. I definitely think its worth trying out and I have created a new branch to test it [here](https://github.com/ml5js/ml5-next-gen/tree/timeseries-as-task). 

However, there are some changes this method necessitates: I have tried implementing DIY time series and DIY neural network classes and the factory functions in the neural network factory function so it is all under one neuralNetwork call. However since the DIY time series class depends on an import of the DIY neural network class, it is not possible to call the DIY time series class inside of the neural network factory function as it needs to import the DIY ts class in the neural network file. This causes a circular dependency as imports of the other files are present. Thus, in the initial commit on the branch I provided, removed the neuralNetwork factory function from the neural network file and moved it all the way to the time series file to prevent this conflict. So far, all of the examples have worked with this method. 

I have yet to try another method where there is a separate file for the neural network factory function, so there is no circular dependencies, however I am not sure where in the file system to put it in and a good naming convention for it. Let me know if you need any more clarification for this!

mop9047 commented on examples/timeSeries-load-model-hand-gestures/model/model_meta.json:
Done :)

gohai commented on src/NeuralNetwork/Sequential/index.js:
```diff
@@ -0,0 +1,210 @@
+    let taskConditions = task;
+    switch (taskConditions) {
+      case "sequenceRegression":
+        layers = tsLayers.regression;
```
Best to keep the name in tsLayers exactly the same as the task, so that searching for e.g. "sequenceRegression" will quickly bring up the layer definition

mop9047 replied:
Done :)

gohai commented on examples/neuralNetwork-sequence-hand-gestures/sketch.js:
```diff
@@ -0,0 +1,180 @@
+let counts = { "Gesture #1": 0, "Gesture #2": 0 };
+
+let state = "collection";
+let currGesture = gestures[0]; //set currGesture to gesture 1 by default
```
(OCD-fueled comment:) The comments in your code seem to oscillate between capitalizing the first word, and not doing it, without much apparent reason. Best to figure out _a_ rule, and stick to it ;)
(And always a space after `//`, please)

PimTournaye commented:
Hi all, I was talking with @gohai about ml5 yesterday, and the subject of audio came up. It was mentioned that this PR was being discussed. There are other libraries out there that can help with audio and machine learning, but it would be nice to also have this capability in ml5, which currently supports pre-trained audio models but doesn't allow users to train custom audio classifiers.

Should audio be considered within this time-series implementation, or would it warrant a separate audio-specific module, or even an adapter to something like [essentia.js](https://github.com/MTG/essentia.js)?

gohai commented on examples/neuralNetwork-sequence-weather-prediction/index.html:
```diff
@@ -0,0 +1,33 @@
+  Learn more about the ml5.js project: https://ml5js.org/
+  ml5.js license and Code of Conduct: https://github.com/ml5js/ml5-next-gen/blob/main/LICENSE.md
+ 
+  This example demonstrates Training a Weather Predictor through ml5.neuralNetwork with sequeceRegression Task.
```
@mop9047 I enjoyed the simplicity of the example. For making the visual output more interesting: how about a bar in a bar-graph gets drawn per hour predicted? (Since you know the training data, you wouldn't need to deal with automatically scaling the y-axis, and instead used some fixed min/max.)
You could divide the width of the canvas into 24 equally-spaced columns, and draw one by one, as the user predicts one hour after the next.

gohai commented on examples/neuralNetwork-sequence-train-mouse-gesture RDP/sketch.js:
```diff
@@ -0,0 +1,132 @@
+ * Learn more about the ml5.js project: https://ml5js.org/
+ * ml5.js license and Code of Conduct: https://github.com/ml5js/ml5-next-gen/blob/main/LICENSE.md
+ *
+ * This example demonstrates How to train your own mouse gesture classifier through ml5.neuralNetwork with sequeceClassification Task.
```
@mop9047 The `train` in the directory name seems superfluous (it's quite long already), and I'd use a dash before `RDP` (since spaces in filenames are at times an issue). Perhaps `neuralNetwork-sequence-mouse-gesture-rdp`?

mop9047 replied:
Yes I agree, done :)

gohai commented on examples/neuralNetwork-sequence-weather-prediction/index.html:
```diff
@@ -0,0 +1,33 @@
+  </body>
+
+  <style>
+    button {
```
I'd keep whatever the default styling for buttons is

mop9047 replied:
Done :)

gohai commented on examples/neuralNetwork-sequence-train-mouse-gesture RDP/index.html:
```diff
@@ -0,0 +1,37 @@
+    <div>
+      <button id="recCircle">Record Circle</button>
+      <button id="recSquare">Record Square</button>
+      <button id="trainBut">Train & Predict</button>
```
"Finished"?

gohai commented on examples/neuralNetwork-sequence-train-mouse-gesture RDP/sketch.js:
```diff
@@ -0,0 +1,132 @@
+    currShape = "circle";
+
+    background(220);
+    text(state + " : " + currShape, 50, 50);
```
It'd be nice if this text already showed up when the sketch launched. "Currently recording: "?

gohai commented on examples/neuralNetwork-sequence-train-mouse-gesture RDP/sketch.js:
```diff
@@ -0,0 +1,132 @@
+    // if state is collection, add whole sequence as X, and shape as Y
+    if (state == "collection") {
+      let target = { label: currShape };
+      let paddedCoordinates = model.padCoordinates(sequence, targetLength);
```
`padCoordinates` runs RDP internally, right? Perhaps you could think of alternative names for this method as options... ("padding" sounds a bit like there's only things added, but if I understand RDP correctly, this will also simplify/decimate if needed?)

Since the algorithm is quite interesting in itself, you might also consider running this continuously, e.g. inside `draw()` and visualize the resulting coordinates it comes up with (e.g. by drawing small squares around the points). That might look quite interesting, and tell our audience a bit more about what's going on.

And lastly: does `padCoordinates()` have some expectation about the input data (e.g. does it assume an `x` and `y`?).

mop9047 replied:
Hi @gohai, I changed the name to `setFixedLength` let me know if the naming is ok or if we should think this through more

gohai commented on examples/neuralNetwork-sequence-load-model-hand-gestures/index.html:
```diff
@@ -0,0 +1,45 @@
+    <meta charset="UTF-8" />
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>ml5.js Time Series Hand Gesture load model</title>
```
I would flip the filename, so that it's clear that this is a further exploration of `hand-gestures`. So: `neuralNetwork-sequence-hand-gestures-load-model`

mop9047 replied:
Done :)

gohai commented on examples/neuralNetwork-sequence-hand-gestures/index.html:
```diff
@@ -0,0 +1,42 @@
+<!--
```
Nitpick: `hand-gestures` is plural while `mouse-gesture` is singular. (Maybe singular throughout?)

gohai commented on examples/neuralNetwork-sequence-load-model-hand-gestures/index.html:
```diff
@@ -0,0 +1,45 @@
+    <script src="sketch.js"></script>
+    <div id="canvasDiv"></div>
+    <p>
+      This example loads a model that is trained with ASL hand gestures for
```
The link to the instructions how to do Hello and Goodbye in ASL (below) is a fantastic touch, but maybe we wouldn't need the text above (including the introduction) ideally?

mop9047 replied:
I did the same for the train model from earlier, hopefully it would be intuitive enough to understand.

gohai commented on examples/neuralNetwork-sequence-train-mouse-gesture RDP/index.html:
```diff
@@ -0,0 +1,37 @@
+    </div>
+
+    <div>
+      <button id="recCircle">Record Circle</button>
```
Or "Start collecting circles"

mop9047 commented on src/NeuralNetwork/Sequential/timeSeriesData.js:
```diff
@@ -0,0 +1,362 @@
+
+import NeuralNetworkData from "../NeuralNetworkData";
+
+class TimeSeriesData extends NeuralNetworkData {
```
Done :)

mop9047 commented on examples/neuralNetwork-sequence-hand-gestures/sketch.js:
```diff
@@ -0,0 +1,180 @@
+  canvas.parent("canvasDiv");
+
+  // setup video capture
+  video = createCapture(VIDEO);
```
yes I agree, done. Also put a comment for the flipped argument in hand pose.

mop9047 commented on examples/neuralNetwork-sequence-hand-gestures/index.html:
```diff
@@ -0,0 +1,42 @@
+    <div id="canvasDiv"></div>
+
+    <p>
+      Instructions: <br />
```
For simplicity sake, I've used the rdp algorithm so it can be any length and no need for the length settings.

I also tried to implement your suggestion but with helpful text. It looks like:

- if no hands in the frame, then the text says put up hands in frame
- if hands are in frame, then it says for which gesture it is recording for
- one training is done, it says put hands in frame again
- if hands are in frame, is says predicting...
- shows prediction once done with

gohai commented on examples/neuralNetwork-sequence-weather-prediction/sketch.js:
```diff
@@ -0,0 +1,155 @@
+ *
+ * This example demonstrates loading JSON data and training a Weather Predictor
+ * through ml5.neuralNetwork with the sequenceRegression task.
+ * XXX: let us know what this dataset is from/about
```
Hi @mop9047. I started cleaning up this example (please make sure to pull the branch before you do future work). Highlighting the parts that need some work here. I'll look at other examples once we're done with this one. 🙌

gohai replied:
Wouldn't it be nicer to look at data of an _actual_ place, one that means something to you? Why not share with us the weather of your home town, or from Brooklyn, or ...? (Should be trivial, but I'll send you a script for US weather stations via email.)

gohai replied:
https://gist.github.com/gohai/dd7d03fb79160817ade763aab33c316c

mop9047 replied:
Hi @gohai I agree that would definitely be nicer. I am trying the data from the script you provided but the precipitation is the whole period so the model auto deletes it (I had no idea it had that feature). Can you help me find the list of the weather station id's? I'll try to find a place with a moving precipitation value

gohai replied:
@mop9047 You could try using the OpenWeatherMap API, which returns you historical data for any `lat`/`long` pair (on the globe). It needs an API key, but supposedly there are a number of free requests per day if you sign up.

```
https://api.openweathermap.org/data/2.5/onecall/timemachine?lat={lat}&lon={lon}&dt={time}&appid={API key}&units=metric
```

Maybe you find some place that has daily rain. Equatorial Rainforests perhaps? (Doesn't this also happen in Indonesia?)

mop9047 replied:
@gohai Not sure if this happens in indonesia, but I did try the monsoon season from where I live in the Philippines and the data just ends up stabilizing at one value. I'm trying to find a place or time that has like a cyclical rain pattern so the predictions look dynamic.

Also in case that you'd want to use it too, I was using a different api in this[ link](https://www.visualcrossing.com/weather-query-builder/#) open weather map is also good but the free tier only offers one piece of data, the one I linked, you can choose as many datapoints as you want.

gohai replied:
Thank you for exploring this!

It's an interesting question: what to do with an example, if it only works with synthetic (toy) data... on one hand it will convey the promise and application of the technique - on the other hand, users will probably try to use it for their own applications/data and then have disappointing results... maybe @shiffman has some thought on this (whenever he has an opportunity to look into this).

I'd love to keep this (or a similar) example around, since it is the only one that actually uses LSTM (rather than CNN). Could you perhaps share the real-life training data you tried with somewhere (gist?), so that we could also play with it?

shiffman replied:
Yes, I agree that we would ideally use a real data set, if it's synthetic I might suggest actually using a math formula and applying randomness to it and then having clear comments and info about how the data was generated? I'm happy to help with gathering data from an API, I'm doing APIs in my class this fall! I love the idea of the data being tied to where you are from @mop9047!

gohai commented on examples/neuralNetwork-sequence-weather-prediction/sketch.js:
```diff
@@ -0,0 +1,155 @@
+  // XXX: - the second and the third parameter need to be the same - just pass one then?
+  // XXX: - the last parameter is 10, but the result has 14 entries 🤷‍♂️ (data has 24)
+  // XXX: - we use the result for inputs and outputs, how about naming those
+  // XXX:   properties with those words directly (rather than "sequences", "targets")
```
@mop9047

gohai commented:
Hi @mop9047!

I tried open-coding the sampling mechanism in the example, see 790a1627a8b12160c75a3098b6d05712e29cb431. Please have a look and tell me if you think this approach is feasible. It also doesn't work anymore (something related to coordinates in `sequentialData.js:188`) - hope you can help me fix it.

* It'd be slightly more ergonomic if your class could simply ignore properties that aren't numbers - then `date` wouldn't need to be unset? 🤔
* Also: aren't we providing the property names for `inputs` and `outputs` in the constructor already? Couldn't it just pick the ones specified? (in which case there'd be even less need to do deep copies of objects to mutate their properties between input and output)

mop9047 commented on examples/neuralNetwork-sequence-weather-prediction/sketch.js:
```diff
@@ -0,0 +1,155 @@
+
+    // XXX: the colors are a bit unintuitive (blue is normally associated
+    // with more rain rather than less) - maybe you could do some research
+    // and use lerpColor() here?
```
Done :)

gohai replied:
Looks great now!

gohai commented:
> Hi @gohai ! I fixed the error, it was because the output also had a length of 10, which made it seq2seq, when the sequenceNN only supports seq2one for now. I changed it so that it takes the multiple samples for the input then take one sample after for the output

Thanks for fixing my problem, @mop9047 - it works well now!

> > * It'd be slightly more ergonomic if your class could simply ignore properties that aren't numbers - then `date` wouldn't need to be unset? 🤔
> Can you clarify what you mean further? when you mean the class do you mean the sequenceNN in general? For now, I believe strings are accepted as inputs since they are also one-hot encoded. It could be useful to have categorical strings in some training use cases.

I see. In this case it's fair to leave them then.

> > * Also: aren't we providing the property names for `inputs` and `outputs` in the constructor already? Couldn't it just pick the ones specified? (in which case there'd be even less need to do deep copies of objects to mutate their properties between input and output)
> For now I think it does throw an error claiming that what is put into the model does not follow the input/output structure defined by the user.

I see some practical upsides in changing it to be "friendly" rather than strict - let me explain:

The user is already defining that his/her inputs shall be `["foo", "bar"]` and the outputs `["foo", "bar"]`. Now if they are passing an _object_ to `nn.addData()` that has some superfluous property `baz`, we could just only pull out values for `foo` and `bar` and leave `baz` alone (without raising an error).

This makes it so that the user wouldn't often need to _clone_ the object in order to achieve what they would want. (The semantics around reference types and shallow vs deep cloning are not always super intuitive for everyone. We'd sidestep this topic by putting more smarts into `addData`.)

gohai commented on src/NeuralNetwork/Sequential/sequentialUtils.js:
```diff
@@ -0,0 +1,764 @@
+  convertValueSequence(xInputs, options = null) {
+    const { inputLabels } = options;
+    if (xInputs.length % inputLabels.length !== 0) {
+      throw new error(
```
@mop9047 This wants to be upper-case. (Got "error is not defined" when I accidentally passed in an empty sequence.)

mop9047 replied:
Thanks for the catch! Have fixed it now, all the other errors in this file are also not capitalized.

gohai commented on examples/neuralNetwork-sequence-mouse-gesture-rdp/sketch.js:
```diff
@@ -0,0 +1,126 @@
+  // For more information about RDP, see:
+  // https://www.youtube.com/watch?v=ZCXkvwLxBrA
+
+  let rdp = model.setFixedLength(sequence, targetLength);
```
@shiffman @MOQN Wondering if you have some (additional) ideas for how to name this (helper) method, which takes a sequence of indeterminate length and returns a sequence of fixed length - using the RDP algorithm?

gohai replied:
`simplify`? `rdp`? `resample`?

gohai commented on examples/neuralNetwork-sequence-hand-gesture/sketch.js:
```diff
@@ -0,0 +1,198 @@
+/*
```
@mop9047 I reworked this example a little bit, please have a look and let me know if you'd rather do things differently!

mop9047 replied:
@gohai Thanks! I think it looks great. The way I designed these sketches was mainly based off of the rock paper scissors example from the NeuralNetwork examples. I personally don't mind straying away from that and I think its also easier without the extra html things in the sketch!

gohai commented on examples/neuralNetwork-sequence-hand-gesture/sketch.js:
```diff
@@ -0,0 +1,198 @@
+    // For more information about RDP, see:
+    // https://www.youtube.com/watch?v=ZCXkvwLxBrA
+
+    let rdp = model.setFixedLength(sequence, targetLength);
```
@mop9047 How is `setFixedLength` dealing with the inputs here? (For an array of numbers, are they implicitly assumed to be x1, y1, x2, y2, etc.?)

mop9047 replied:
@gohai yes this is an assumption that I made based on other sample code using the hand pose model. Would there be any case where this would cause any issues?

gohai replied:
No this makes sense. Will just need good documentation ultimately!

gohai commented on examples/neuralNetwork-sequence-hand-gesture/sketch.js:
```diff
@@ -0,0 +1,198 @@
+// Return the tracked hand points as flattened array of 84 numbers
+// for use as input to the neural network
+
+function getKeypoints(whichHands = ["Left", "Right"]) {
```
@mop9047 This was a bit tricky for one reason: handPose can return one or two hands, which means this function would return _either_ 42 or 84 numbers. (I am not sure if the order the hands appear in the array is guaranteed either.)

Currently this now returns always 84 numbers (with zeros for hands that aren't showing), but I am not sure if this is the best way... do you think it makes sense for this to be able to record gestures that make use of both hands? Or, other direction: would it be bad to specify that a specific hand need to be used to make a gesture (and the sketch only looks for that one)?

mop9047 replied:
From my own experience, it depends on how complex and how many categories are, in our case of two categories and only some simple gestures we definitely can do lesser points but often in sign language as an example there are similar gestures that from the perspective of the model will look the same.

gohai commented on examples/neuralNetwork-sequence-hand-gesture-load-model/index.html:
```diff
@@ -0,0 +1,28 @@
+    <meta charset="UTF-8" />
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>ml5.js neuralNetwork Hand Gesture Loading Pre-trained Model Example</title>
```
@mop9047 I slightly reworked the last example - please have a look if anything is amiss 🥳

gohai commented on examples/neuralNetwork-sequence-hand-gesture-load-model/sketch.js:
```diff
@@ -0,0 +1,158 @@
+let isModelLoaded = false;
+
+let sequence = [];
+let sequenceLength = 50;
```
@mop9047 Change this to 30? If you can, please also upload re-trained model files (with 84 dimensions)

gohai replied:
That's great - thank you!

I think both (showing or not showing the path in this example) could work. Perhaps leaning a bit towards the latter, so that there is more differentiation to what your other hand-gesture example offers, and we keep this a way of demonstrating how a clean, "user-facing" application could look like?

gohai commented on src/NeuralNetwork/Sequential/index.js:
```diff
@@ -0,0 +1,215 @@
+
+// call an extension of DIY Neural Network as a new class, override select methods
+// which are seen below:
+class DIYSequential extends DiyNeuralNetwork {
```
@mop9047 Perhaps as the last task for this period: please try to add some friendly error messages where possible 🙌 The few instances I have encountered:

* training with only one (output) class: `Error in oneHot: depth must be >=2, but it is 1
    at oneHot_` (if this is worth fixing in the regular NN, please open a separate PR!)
* forgetting to call `normalizeData`: `nput 0 is incompatible with layer conv1d_Conv1D1: expected ndim=3, found ndim=2
    at Conv1D.assertInputCompatibility`
* varying the length of the sequence during training: `Error when checking input: expected conv1d_Conv1D1_input to have 3 dimension(s). but got array with shape 0,1
    at standardizeInputData`
* during inference, submitting sequences with length 30 against a model that was trained with 50: `Error when checking : expected conv1d_Conv1D1_input to have shape [null,50,40] but got array with shape [1,30,42].`

mop9047 replied:
@gohai I have created the some friendlier errors for these cases, let me know if the wording of the error is a bit confusing

mop9047 replied:
For the first error, it also happens in the base neuralNetwork class so I made PR for the first error here: https://github.com/ml5js/ml5-next-gen/pull/268

gohai replied:
Great - thanks so much @mop9047!

gohai commented on examples/neuralNetwork-sequence-hand-gesture-load-model/index.html:
```diff
@@ -0,0 +1,28 @@
+<!--
```
@shiffman @MOQN *Very unimportant*, but I thought I raise this here before I forget: was wondering if we prefer `neuralNetwork-sequence-...` or a (shorter) `neuralSequence-...`? (Feeling the former is truer to the hierarchy - but maybe too long?)
